man/openk.1
.Dd November 25, 2024
.Dt OPNK 1
.Os 
.
.Sh NAME
.Nm openk
.Nd universal open (like a punk) for offpunk
.
.Sh SYNOPSIS
.Nm
.Op \-mode Ar MODE
.Op \-linkmode Ar LINKMODE
.Op \-cache-validity Ar CACHE_VALIDITY
.Op Ar INPUT ...
.Nm
.Fl h | \-help
.
.Sh DESCRIPTION
openk is an universal open command tool
that will try to display any file in the pager
.Xr less 1
after rendering its content with
.Xr ansicat 1 .
If that fails,
openk will fallback to opening the file with
.Xr xdg-open 1 .
If given an URL as input instead of a path,
openk will rely on
.Xr netcache 1
to get the networked content.
.Ss Positional arguments
.Bl -tag -width Ds -offset indent
.It Ar INPUT
path to the file or URL to open.
.El
.Ss Keyword arguments
.Bl -tag -width Ds -offset indent
.It Fl h , \-help
Show a help message and exit
.It Fl \-mode Ar MODE
MODE to use to render to choose between normal (default), full or source
.It Fl \-linkmode Ar LINKMODE
LINKMODE to use to render links. Available are
.Ic none
(default) and
.Ic end .
This option will not effect render mode source.
.It Fl \-cache-validity CACHE_VALIDITY
Maximum age (in second) of the cached version before redownloading a new version.
.El
.
.Sh EXIT STATUS
.Ex -std
.
.Sh SEE ALSO
.Xr ansicat 1 ,
.Xr migrate-offpunk-cache 1 ,
.Xr netcache 1 ,
.Xr offpunk 1 ,
.Lk https://offpunk.net/
.
.Sh AUTHORS
.An Lionel Dricot (Ploum) Aq Mt offpunk2 at ploum.eu


CHANGELOG
# Offpunk History

## 3.0-beta3 - unreleased
- added ftr_site_config output to "version"
- added "wgl" and "wde" shortcuts to wikipedia GL & DE (JMCS)

## 3.0-beta2 - February 6th 2026
Changes since 3.0-beta1
- "bugreport" allows to send a bug report to the -devel list
- "theme" now accept "none" as a color to remove part of the theme you don’t want
- symlink opnk.py to openk.py to handle deprecation
- BUGFIX: history now consider multiple successive views of a page as one visit.
- BUGFIX: do not urlify mailto, gopher and local urls
- BUGFIX: In "openk", self.last_width was badly reinitialized after cleanup()
- CRASHFIX: in "ansicat", build_body_and_links was crashing with empty links
- CRASHFIX: "openk" was crashing when used in a shell pipe
- improve Gopher selector code to be more robust (by JMCS)
- TRANSLATION: Spanish - ES (by JMCS)
- TRANSLATION: Galician - GC (by JMCS)

## 3.0-beta1 - February 2nd 2026
This is a beta release before 3.0. Its goal is to find bugs, allow for a 
"string freeze" period for translators and give packagers time to adapt 
to the major changes: adding translations to their packages, adding 
xkcdpunk and unmerdify. Packagers are advised to build the package to 
report bugs but *not* to upload to their respective stable 
distribution.
Beta-testers and documenters are welcome to test and document new 
features and new tools.
Important changes:
- "opnk" is deprecated and replaced by "openk"
- Default image size now 100 instead of 40 (can be restored with "set images_size 40")
- Links to available feeds are now displayed at the bottom of HTML pages
- "root" has been modified to go to the root of the capsule/userspace "root /" for real root
- Images are now displayed in Gemini. This can be disabled with "set gemini_images false"
- "ls" command is deprecated
Features:
- Offpunk is now translatable (by JMCS)
- new "xkcdpunk" command-line tool to directly access your XKCD comics
- new "websearch" command which default to wiby.me
- new "share" command to send a page by email
- new "reply" command to send an email to the author of the page
- "cookies" command and http cookies are now supported in netcache (by Urja)
- "set default_cmd" now allows to configure what happen on empty lines
- "view switch" changes between readable/full view (by Andrew Fowlie)
- customize prompt with "set prompt_on" and "set prompt_off" (by Andrew Fowlie)
- "help help" now sends an email to the offpunk-users mailing-list
- "links" command now display all links
- new "blocked_link" object in theme. By default, blocked link are in red.
- new "theme preset" to allow switching between multiple hardcoded themes like "yellow" or "cyan"
Unmerdify alpha support (by Vincent Jousse):
- Add "unmerdify" tool to extract content from HTML page
- if "ftr_site_config" is set, unmerdify will be enabled in "ansicat"
- ansicat will fallback to readability if unmerdify fails
- "info" now explain how the content was cleaned
Others:
- In "lists", only consider files ending with .gmi (bug #46 by woffs)
- support for new links color in gopher (by JMCS)
- fix an off by one bug in "cp url XX" 
- Ansicat: Fix incorrectly identifying RSS feeds as XHTML when there was no entry
- Fix sync hanging on gopher interactive prompt (by JMCS)
- Fix crash when trying to parse HTML without BS4 (reported by Hoël Bézier)
- Automatically replace space by %20 in img url in HTML
- Ansicat: HTML Render img with data-src if src doesn’t link to a valid image
- Opnk: fix a crash when trying to cache-clean uncached url (such as images)
- added mime-type "application/pgp-keys" as to be opened by TextRenderer
- "redirect" has been refactored to acts on the netcache level
- ansicat: in HTML, the <a> link to a picture is only displayed if different from the picture itself

## 2.8 - November 13th 2025
- New option "images_size" to use with "set" (idea by aelius)
- Automatically use newly created certificates in Gemini (by JMCS)
- Fix a rare crash when listing urls in a non-existant page
- ansicat: Trying to get render preformatted text correctly in HTML
- ansicat: html <code> should be inline, not a separate block like <pre>
- ansicat: <abbr> is rendered as <code>
- fix "up" not working correctly on gopher (thanks JMCS)
- gopher: implement support for item type 7 (by JMCS)
- fix out of range index on page without links (by Alexander W. Jans)
- User-agent now includes URL to avoid blocking (suggested by cks)
- Adding gemtext rendering to message/rfc822 files 
- ansicat: support for <base> in html rendering
- gopher: consider .xml files as feeds (by JMCS)
- gopher: detect image types in gopher to open them internally (by JMCS)
- "reload" now delete cached rendering, online and offline
- ansicat: remove soft-hyphens when rendering any content

## 2.7.1 - April 9th 2025 - The "2.7 was not (completely) a joke" release
See 2.7 changelog. This was "real" release ;-)
- Reverted the tutorial back to Offpunk. The name was a joke but the rest of the release is real.
- added "application/xhtml+xml" mimetype to be rendered as html
- according to html spec, <pre> content should be parsed as HTML.

## 2.7 - April 1st 2025 - XKCDpunk release
APRIL FOOL: Offpunk is now renamed to XKCDpunk. See
https://ploum.net/2025-04-01-xkcdpunk.html
- Introducing command "xkcd" to display a given xkcd comic.
- (April fool) Switching website from offpunk.net to xkcdpunk.net
- blocklist: marked x.com, twitter.com and youtube.com as blocked to save bandwith until we have a viable option to use them with offpunk. We plan to block everything but xkcd.com.
CHANGES:
- "shell" (or "!") now works even without content (so you can run it on startup)
- offpunk: improved "version" to help debugging on the mailing-list
FIXES:
- "reload" on a too large content will actually fetch it
- Consider mimetype "message/news" as "gemtext" because file is sometimes confused. (reported by JMCS)
- ignore annoying warnings for LibreSSL users (by Anna cyberTailor)

openk.py
#!/usr/bin/env python3
# openk stand for "Open like a PuNK".
# It will open any file or URL and display it nicely in less.
# If not possible, it will fallback to xdg-open
# URL are retrieved through netcache
import argparse
import fnmatch
import os
import shutil
import sys
import tempfile
import time
import gettext

import ansicat
import netcache
import offutils
import offblocklist
from offutils import (
        GREPCMD, 
        is_local, 
        mode_url, 
        run, 
        term_width, 
        unmode_url, 
        init_config,
        send_email,
        _HAS_XDGOPEN,
        _LOCALE_DIR,
        )

gettext.bindtextdomain('offpunk', _LOCALE_DIR)
gettext.textdomain('offpunk')
_ = gettext.gettext

less_version = 0
if not shutil.which("less"):
    print(_('Please install the pager "less" to run Offpunk.'))
    print(_("If you wish to use another pager, send me an email !"))
    print(
        _('(I’m really curious to hear about people not having "less" on their system.)')
    )
    sys.exit()
output = run("less --version")
# We get less Version (which is the only integer on the first line)
words = output.split("\n")[0].split()
less_version = 0
for w in words:
    # On macOS the version can be something like 581.2 not just an int:
    if all(_.isdigit() for _ in w.split(".")):
        less_version = int(w.split(".", 1)[0])
# restoring position only works for version of less > 572
if less_version >= 572:
    _LESS_RESTORE_POSITION = True
else:
    _LESS_RESTORE_POSITION = False


# _DEFAULT_LESS = "less -EXFRfM -PMurl\ lines\ \%lt-\%lb/\%L\ \%Pb\%$ %s"
# -E : quit when reaching end of file (to behave like "cat")
# -F : quit if content fits the screen (behave like "cat")
# -X : does not clear the screen
# -R : interpret ANSI colors correctly
# -f : suppress warning for some contents
# -M : long prompt (to have info about where you are in the file)
# -W : hilite the new first line after a page skip (space)
# -i : ignore case in search
# -S : do not wrap long lines. Wrapping is done by offpunk, longlines
# are there on purpose (surch in asciiart)
# --incsearch : incremental search starting rev581
def less_cmd(file, histfile=None, cat=False, grep=None):
    less_prompt = "page %%d/%%D- lines %%lb/%%L - %%Pb\\%%"
    if less_version >= 581:
        less_base = 'less --incsearch --save-marks -~ -XRfWiS -P "%s"' % less_prompt
    elif less_version >= 572:
        less_base = "less --save-marks -XRfMWiS"
    else:
        less_base = "less -XRfMWiS"
    _DEFAULT_LESS = less_base + " \"+''\" %s"
    _DEFAULT_CAT = less_base + " -EF %s"
    if histfile:
        env = {"LESSHISTFILE": histfile}
    else:
        env = {}
    if cat and not grep:
        cmd_str = _DEFAULT_CAT
    elif grep:
        grep_cmd = GREPCMD
        # case insensitive for lowercase search
        if grep.islower():
            grep_cmd += " -i"
        cmd_str = _DEFAULT_CAT + "|" + grep_cmd + " %s" % grep
    else:
        cmd_str = _DEFAULT_LESS
    run(cmd_str, parameter=file, direct_output=True, env=env)


class opencache:
    def __init__(self):
        # We have a cache of the rendering of file and, for each one,
        # a less_histfile containing the current position in the file
        self.temp_files = {}
        self.less_histfile = {}
        # This dictionary contains an url -> ansirenderer mapping. This allows
        # to reuse a renderer when visiting several times the same URL during
        # the same session
        # We save the time at which the renderer was created in renderer_time
        # This way, we can invalidate the renderer if a new version of the source
        # has been downloaded
        self.rendererdic = {}
        self.renderer_time = {}
        self.mime_handlers = {}
        self.last_mode = {}
        self.last_width = term_width(absolute=True)
        self.redirects = offblocklist.redirects

    def _get_handler_cmd(self, mimetype,file_extension=None):
        # Now look for a handler for this mimetype
        # Consider exact matches before wildcard matches
        exact_matches = []
        wildcard_matches = []
        for handled_mime, cmd_str in self.mime_handlers.items():
            if "*" in handled_mime:
                wildcard_matches.append((handled_mime, cmd_str))
            else:
                exact_matches.append((handled_mime, cmd_str))
        for handled_mime, cmd_str in exact_matches + wildcard_matches:
            if fnmatch.fnmatch(mimetype, handled_mime):
                break
            #we try to match the file extension, with a starting dot or not
            elif file_extension == handled_mime.strip("."): 
                break
        else:
            # Use "xdg-open" as a last resort.
            if _HAS_XDGOPEN:
                cmd_str = "xdg-open %s"
            else:
                #TRANSLATORS: keep echo and %s, translate the text between ""
                cmd_str = _('echo "Can’t find how to open "%s')
                print(_("Please install xdg-open (usually from xdg-util package)"))
        return cmd_str

    # Return the handler for a specific mimetype.
    # Return the whole dic if no specific mime provided
    def get_handlers(self, mime=None):
        if mime and mime in self.mime_handlers.keys():
            return self.mime_handlers[mime]
        elif mime:
            return None
        else:
            return self.mime_handlers

    def set_handler(self, mime, handler):
        if "%s" not in handler:
            #if no %s, we automatically add one. I can’t think of any usecase
            # where it should not be part of the handler
            handler += " %s"
        previous = None
        if mime in self.mime_handlers.keys():
            previous = self.mime_handlers[mime]
        self.mime_handlers[mime] = handler

    def get_renderer(self, inpath, mode=None, theme=None,**kwargs):
        # We remove the ##offpunk_mode= from the URL
        # If mode is already set, we don’t use the part from the URL
        inpath, newmode = unmode_url(inpath)
        if not mode:
            mode = newmode
        # If we still doesn’t have a mode, we see if we used one before
        if not mode and inpath in self.last_mode.keys():
            mode = self.last_mode[inpath]
        elif not mode:
            # default mode is readable
            mode = "readable"
        renderer = None
        path = netcache.get_cache_path(inpath)
        if path:
            usecache = inpath in self.rendererdic.keys() and not is_local(inpath)
            # Screen size may have changed
            width = term_width(absolute=True)
            if usecache and self.last_width != width:
                self.cleanup()
                usecache = False
            if usecache:
                if inpath in self.renderer_time.keys():
                    last_downloaded = netcache.cache_last_modified(inpath)
                    last_cached = self.renderer_time[inpath]
                    if last_cached and last_downloaded:
                        usecache = last_cached > last_downloaded
                    else:
                        usecache = False
                else:
                    usecache = False
            if not usecache:
                renderer = ansicat.renderer_from_file(path, url=inpath, theme=theme,\
                                                      redirectlist=self.redirects,**kwargs)
                if renderer:
                    self.rendererdic[inpath] = renderer
                    self.renderer_time[inpath] = int(time.time())
            else:
                renderer = self.rendererdic[inpath]
        return renderer

    def get_temp_filename(self, url):
        if url in self.temp_files.keys():
            return self.temp_files[url]
        else:
            return None

    def openk(self, inpath, mode="readable", terminal=True, grep=None, theme=None, \
                link=None, direct_open_unsupported=False, **kwargs):
        # Return True if inpath opened in Terminal
        # False otherwise
        # also returns the url in case it has been modified
        # if terminal = False, we don’t try to open in the terminal,
        # we immediately fallback to xdg-open.
        # netcache currently provide the path if it’s a file.
        # If link is a digit, we open that link number instead of the inpath
        # If direct_open_unsupported, we don’t print the "unsupported warning"
        # and, instead, immediately fallback to external open
        if not offutils.is_local(inpath):
            if mode:
                kwargs["images_mode"] = mode
            cachepath, inpath = netcache.fetch(inpath, redirects=self.redirects,**kwargs)
            if not cachepath:
                return False, inpath
        # folowing line is for :// which are locals (file,list)
        elif "://" in inpath:
            cachepath, inpath = netcache.fetch(inpath, redirects=self.redirects,**kwargs)
        elif inpath.startswith("mailto:"):
            cachepath = inpath
        elif os.path.exists(inpath):
            cachepath = inpath
        else:
            print(_("%s does not exist") % inpath)
            return False, inpath
        renderer = self.get_renderer(inpath, mode=mode, theme=theme, **kwargs)
        if link and link.isdigit():
            inpath = renderer.get_link(int(link)) 
            renderer = self.get_renderer(inpath, mode=mode, theme=theme, **kwargs)
        if renderer and mode:
            renderer.set_mode(mode)
            self.last_mode[inpath] = mode
        if not mode and inpath in self.last_mode.keys():
            mode = self.last_mode[inpath]
            renderer.set_mode(mode)
        # we use the full moded url as key for the dictionary
        key = mode_url(inpath, mode)
        if renderer and not renderer.is_format_supported() and direct_open_unsupported:
            terminal = False
        if terminal and renderer:
            # If this is an image and we have chafa/timg, we
            # don’t use less, we call it directly
            if renderer.has_direct_display():
                renderer.display(mode=mode, directdisplay=True)
                return True, inpath
            else:
                body = renderer.display(mode=mode)
                # Should we use the cache ? only if it is not local and there’s a cache
                usecache = key in self.temp_files and not is_local(inpath)
                if usecache:
                    # and the cache is still valid!
                    last_downloaded = netcache.cache_last_modified(inpath)
                    last_cached = os.path.getmtime(self.temp_files[key])
                    if last_downloaded > last_cached:
                        usecache = False
                        self.temp_files.pop(key)
                        self.less_histfile.pop(key)
                # We actually put the body in a tmpfile before giving it to less
                if not usecache:
                    tmpf = tempfile.NamedTemporaryFile(
                        "w", encoding="UTF-8", delete=False
                    )
                    self.temp_files[key] = tmpf.name
                    tmpf.write(body)
                    tmpf.close()
                if key not in self.less_histfile:
                    firsttime = True
                    tmpf = tempfile.NamedTemporaryFile(
                        "w", encoding="UTF-8", delete=False
                    )
                    self.less_histfile[key] = tmpf.name
                else:
                    # We don’t want to restore positions in lists
                    firsttime = is_local(inpath)
                less_cmd(
                    self.temp_files[key],
                    histfile=self.less_histfile[key],
                    cat=firsttime,
                    grep=grep,
                )
                return True, inpath
        # maybe, we have no renderer. Or we want to skip it.
        else:
            mimetype = ansicat.get_mime(cachepath)
            #we find the file extension by taking the last part of the path
            #and finding a dot.
            last_part = cachepath.split("/")[-1]
            extension = None
            if last_part and "." in last_part:
                extension = last_part.split(".")[-1]
            if mimetype == "mailto":
                mail = inpath[7:]
                send_email(mail,toconfirm=True)
                return False, inpath
            else:
                cmd_str = self._get_handler_cmd(mimetype,file_extension=extension)
            #TRANSLATORS translate only "MY_PREFERED_APP"
            change_cmd = _("\"handler %s MY_PREFERED_APP %%s\"")%mimetype
            try:
                #we don’t write the info if directly opening to avoid 
                #being verbose in openk
                if not direct_open_unsupported:
                    print(_("External open of type %s with \"%s\"")%(mimetype,cmd_str))
                    print(_("You can change the default handler with %s")%change_cmd)
                run(
                    cmd_str,
                    parameter=netcache.get_cache_path(inpath),
                    direct_output=True,
                )
                return True, inpath
            except FileNotFoundError:
                print(_("Handler program %s not found!") % shlex.split(cmd_str)[0])
                print(_("You can use the ! command to specify another handler program\
                        or pipeline."))

                print(_("You can change the default handler with %s")%change_cmd)
            return False, inpath

    # We remove the renderers from the cache and we also delete temp files
    def cleanup(self):
        while len(self.temp_files) > 0:
            os.remove(self.temp_files.popitem()[1])
        while len(self.less_histfile) > 0:
            os.remove(self.less_histfile.popitem()[1])
        #After cleanup, we set the current size of the terminal
        self.last_width = term_width(absolute=True)
        self.rendererdic = {}
        self.renderer_time = {}
        self.last_mode = {}

    # Clean only a specificif url cache
    def clean_url(self,url,mode=None):
        def delfile(path):
            if os.path.isfile(path):
                os.remove(path)
        # First, we take the full URL
        if mode:
            url = mode_url(url,mode)
        # Unrendered element, such as picture, are not in the dictionnaries
        if url in self.temp_files:
            delfile(self.temp_files.pop(url))
        if url in self.less_histfile:
            delfile(self.less_histfile.pop(url))
        url, newmode = unmode_url(url)
        if url in self.rendererdic:
            self.rendererdic.pop(url)
        if url in self.renderer_time:
            self.renderer_time.pop(url)


def main():
    descri = _("openk is an universal open command tool that will try to display any file \
             in the pager less after rendering its content with ansicat. If that fails, \
             openk will fallback to opening the file with xdg-open. If given an URL as input \
             instead of a path, openk will rely on netcache to get the networked content.")
    parser = argparse.ArgumentParser(prog="openk", description=descri)
    parser.add_argument(
        "--mode",
        metavar="MODE",
        help=_("Which mode should be used to render: normal (default), full or source.\
                                With HTML, the normal mode try to extract the article."),
    )
    parser.add_argument(
        "--linkmode",
        choices=[
            "none",
            "end",
        ],
        help=_("Which mode should be used to render links: none (default) or end"),
    )
    parser.add_argument(
        "content",
        metavar="INPUT",
        nargs="*",
        default=sys.stdin,
        help=_("Path to the file or URL to open"),
    )
    parser.add_argument(
        "--cache-validity",
        type=int,
        default=0,
        help=_("maximum age, in second, of the cached version before \
                                redownloading a new version"),
    )
    args = parser.parse_args()
    cache = opencache()
    #we read the startup config and we only care about the "handler" command
    cmds = init_config(skip_go=True,interactive=False,verbose=False)
    for cmd in cmds:
        splitted = cmd.split(maxsplit=2)
        if len(splitted) >= 3 and splitted[0] == "handler":
            cache.set_handler(splitted[1],splitted[2])
    # if the second argument is an integer, we associate it with the previous url
    # to use as a link_id
    if (type(args.content) == list) and len(args.content) == 2 and args.content[1].isdigit():
        url = args.content[0]
        link_id = args.content[1]
        cache.openk(url, mode=args.mode, validity=args.cache_validity, link=link_id,\
                    direct_open_unsupported=True, linkmode=args.linkmode)
    else:
        for f in args.content:
            cache.openk(f, mode=args.mode, validity=args.cache_validity,\
                        direct_open_unsupported=True, linkmode=args.linkmode)

if __name__ == "__main__":
    if "opnk" in sys.argv[0]:
        print("WARNING: opnk.py has been deprecated in favour of openk.py")
        print("*******  Replace all your opnk calls by \"openk\"")
    main()
